pandas/tests/frame/test_apply.py

```diff
@@ -482,6 +486,200 @@ def test_apply_non_numpy_dtype(self):
         assert_frame_equal(result, df)
 
 
+class TestInferOutputShape(object):
+    # the user has supplied an opaque UDF where
+    # they are transforming the input that requires
+    # us to infer the output
+
+    def test_infer_row_shape(self):
+        # gh-17437
+        # if row shape is changing, infer it
+        df = pd.DataFrame(np.random.rand(10, 2))
+        result = df.apply(np.fft.fft, axis=0)
```

Comment:
There is some inconsistency here in how we handle this compared to `axis=1` (but that is certainly something that can be discussed in a follow-up issue)

Reply:
I don't think there IS an inconsistency. We cannot set the column names of the result frame, UNLESS a Series is returned. Sure we *can*, but it would then be inconsistent depending on the length of the returned result (e.g. if it matches in columns we can, but then if it doesn't we cannot set them). Simply better to say that infer does not set column names at all.